### PR TITLE
Sample Linter

### PR DESCRIPTION
**Pull request by johnpierson**
This PR contains the sample linter that has the following rules and documentation in place:

- **Slider as input rule:** This rule alerts the end user that they might want to do so if the slider is not input or renamed. If a slider is not being used as input, it is typically unnecessary and a number node or a code block would suffice.
- **Dropdown as Input Rule:** Similar to the slider rule, but for dropdowns. If you aren't using it as an input, it is worth considering an alternative.
- **Nodes not in Groups:** Having ungrouped nodes is a bad practice. This section allows the linter to flag this behavior based on linter settings in the extra folder in the extension directory.

20230710-SampleLinter

FYI @saintentropy @lillismith

### CODE REVIEW

**Comment by lillismith** [comment]
@twastvedt  This looks good to me from a content perspective. Could you give the code a review?

File: dynamo_viewExtension/Sample View Extension/extra/SampleViewExtension_ViewExtensionDefinition.xml
Comment (twastvedt) [inline dynamo_viewExtension/Sample View Extension/extra/SampleViewExtension_ViewExtensionDefinition.xml:1]: Seems like other xml files have 2 space indentation. I think we shouldn't include this change?

File: src/SampleLinter/SampleLinter.csproj [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    <None Include="packages.config" />
+  </ItemGroup>
+  <ItemGroup />
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Target Name="AfterBuild">
+    <ItemGroup>
+      <PackageDll Include="$(ProjectDir)bin\$(ConfigurationName)\SampleLinter.dll" />
+      <PackageXml Include="$(ProjectDir)bin\$(ConfigurationName)\SampleLinter_ExtensionDefinition.xml" />
+    </ItemGroup>
+    <Copy SourceFiles="@(PackageDll)" DestinationFolder="$(ProjectDir)..\..\dynamo_viewExtension\Sample View Extension\bin" />
```
Comment (mjkkirschner) [inline src/SampleLinter/SampleLinter.csproj]: can you enumerate the structure here, is this being bundled with the existing sample view extension? 
Seems like theres a new extension, a new view extension, and a package?

File: src/SampleLinter/SampleLinter.csproj [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    </Reference>
+    <Reference Include="System.Numerics" />
+    <Reference Include="System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>packages\System.Numerics.Vectors.4.5.0\lib\net46\System.Numerics.Vectors.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>packages\System.Runtime.CompilerServices.Unsafe.4.5.3\lib\net461\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
+    </Reference>
+    <Reference Include="System.Windows.Interactivity, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>packages\Prism.4.1.0.0\lib\NET40\System.Windows.Interactivity.dll</HintPath>
```
Comment (mjkkirschner) [inline src/SampleLinter/SampleLinter.csproj:120]: it looks like the entire `packages` folder has been committed, please remove it.

File: src/SampleLinter/SampleLinter.csproj [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{146EBF48-E7A0-4ABE-809D-D7F3059E4EE1}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>SampleLinter</RootNamespace>
+    <AssemblyName>SampleLinter</AssemblyName>
+    <TargetFrameworkVersion>v4.8.1</TargetFrameworkVersion>
```
Comment (twastvedt) [inline src/SampleLinter/SampleLinter.csproj]: This should be v4.8. `v4.8.1` isn't a valid value for [TargetFrameworkVersion](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-target-framework-and-target-platform?view=vs-2022#target-framework-and-profile), even though it is a .net framework version.

File: src/SampleLinter/SampleLinter.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+            //add each of our rules to the linter
+            _sliderInputRule = new SampleSliderInputLinterRule();
+            AddLinterRule(_sliderInputRule);
+
+            _DropdownRule = new SampleDropdownInputLinterRule();
+            AddLinterRule(_DropdownRule);
+
+            _noGroupsRule = new NoGroupsLinterRule(_linterSettings);
+            AddLinterRule(_noGroupsRule);
```
Comment (twastvedt) [inline src/SampleLinter/SampleLinter.cs:41]: Is there much reason in typical use cases that we would need access to the specific rule objects? If not, maybe it would be useful to demonstrate a more scalable pattern here, with a `HashSet<NodeLinterRule>` property that we can loop through to add and remove the rules? I'm just cringing thinking of the version of this with 20 rules.

File: src/SampleLinter/SampleLinter.csproj
Comment (twastvedt) [inline src/SampleLinter/SampleLinter.csproj:1]: I think if we're publishing this for others to copy and use, it would be worth converting this project to SDK style as you did with the migrator tool. Better dev experience all around.

**Comment by lillismith** [comment]
@johnpierson any thoughts on above review?  Seems like we should address these issues before publishing on the package manager.

**Comment by johnpierson** [comment]
Let me check it out @lillismith . I also do not believe this is going on the package manager though. Dynamo samples traditionally have only been a github repo

**Comment by lillismith** [comment]
Ah, yes, no package manager. You are correct @johnpierson  Still we should wrap up these comments before accepting the PR and the blog post then.

File: src/SampleLinter/SampleLinter.csproj (hunk 1 above)
Comment (johnpierson) [reply src/SampleLinter/SampleLinter.csproj]: Fixed

File: src/SampleLinter/SampleLinter.csproj (hunk 2 above)
Comment (johnpierson) [reply src/SampleLinter/SampleLinter.csproj:120]: Fixed

File: src/SampleLinter/SampleLinter.csproj (hunk 3 above)
Comment (johnpierson) [reply src/SampleLinter/SampleLinter.csproj]: Fixed

File: dynamo_viewExtension/Sample View Extension/extra/SampleViewExtension_ViewExtensionDefinition.xml
Comment (johnpierson) [reply dynamo_viewExtension/Sample View Extension/extra/SampleViewExtension_ViewExtensionDefinition.xml:1]: fixed

**Comment by lillismith** [comment]
@mjkkirschner  looks like @johnpierson  has resolved all the comments, could you advise if we should accept this PR now or if you would like someone else to have a look?  Thank you!

**Comment by johnpierson** [comment]
The last part was regarding Trygve's comment about swapping the rules to a HashSet, I can take a look at that. But part of the reason I did it kind of simply was because this was an example, and I am not super familiar with HashSets and the Linter.

**Review by mjkkirschner — CHANGES_REQUESTED** [review]
the packages folder is still being committed. You can tell because the PR is bringing in 155,224 changes to 203 files.

**Comment by johnpierson** [comment]
Hmm, what is the best way to remove that? Just delete it on my end?

**Comment by mjkkirschner** [comment]
> @mjkkirschner looks like @johnpierson has resolved all the comments, could you advise if we should accept this PR now or if you would like someone else to have a look? Thank you!

Hi @lillismith I am just providing general feedback about commits to this repo in general and best practices for .net/c# development. I have no idea about the linter, how it works, or what a good sample for it looks like.

I think you should have someone review it who is familiar with it. Providing a good sample with good best practices would be ideal for our users, providing a poor sample will lead to a proliferation of poor/ non idiomatic usage.

Who is familiar with this feature, maybe @BogdanZavu ?

**Comment by lillismith** [comment]
@mjkkirschner Thanks very much for taking a look.  @BogdanZavu  could you please review and test the linter and then accept this PR if you agree it is in good shape?  Thank you!

**Comment by BogdanZavu** [comment]
@mjkkirschner @lillismith Yes, I will look into it this week.

File: src/SampleLinter/Rules/NoGroupsLinterRule.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+using Dynamo.Linting.Rules;
+
+namespace SampleLinter.Rules
+{
+    public class NoGroupsLinterRule : GraphLinterRule
+    {
+        public override string Id => "EF17A093-4E94-4AE9-8876-18631396A23A"; //this id is unique to each linter rule
+        public override SeverityCodesEnum SeverityCode => SeverityCodesEnum.Warning;
+        public override string Description => $"There are {_nodesNotInGroups.Count} nodes that are not in groups.";
+        public override string CallToAction => $"You have nodes that are not in groups. Grouping your nodes is a best-practice when authoring Dynamo graphs. {_ungroupedAllowed} nodes are allowed outside of groups based on the Linter Settings.";
```
Comment (BogdanZavu) [inline src/SampleLinter/Rules/NoGroupsLinterRule.cs]: I see that the other samples are localized so I think we'll need to do the same.

File: .gitignore [hunk 6]
```diff
@@ -136,6 +136,7 @@ publish/
 
 src/packages/
 !src/packages/repositories.config
+src/SampleLinter/packages
```
Comment (BogdanZavu) [inline .gitignore]: Looks like all the other samples are putting their package dependencies in src/packages. Let's do the same, it also makes it a little easier for us to track third parties.

File: dynamo_linter/Sample Linter/extra/LinterSettings.xml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinterSettings xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <AllowedUngroupedNodes>5</AllowedUngroupedNodes>
```
Comment (BogdanZavu) [inline dynamo_linter/Sample Linter/extra/LinterSettings.xml:3]: @johnpierson great idea! And perhaps we can expand here a little bit and have an entry for the other rules as well. I guess for the other rules it will be just active or not. This way one has more control over the linter without being a developer.

File: src/SampleLinter/MigrationSettingsView.xaml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿<Window x:Class="SampleLinter.MigrationSettingsView"
```
Comment (BogdanZavu) [inline src/SampleLinter/MigrationSettingsView.xaml]: I'm guessing MigrationSettings is not part of this PR ? Or am I missing something

File: src/SampleLinter/SampleLinter.sln [hunk 9]
```diff
@@ -0,0 +1,25 @@
+﻿
```
Comment (BogdanZavu) [inline src/SampleLinter/SampleLinter.sln]: I guess we don't need SampleLinter.sln since we just need to integrate in DynamoSamples.sln

File: src/SampleLinter/Rules/NoGroupsLinterRule.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Dynamo.Graph.Nodes;
+using Dynamo.Graph.Workspaces;
+using Dynamo.Linting.Interfaces;
+using Dynamo.Linting.Rules;
+
+namespace SampleLinter.Rules
+{
```
Comment (BogdanZavu) [inline src/SampleLinter/Rules/NoGroupsLinterRule.cs:10]: Let's add the rules description from your pr head on top of each rule as a summary comment.

File: src/SampleLinter/Rules/NoGroupsLinterRule.cs (hunk 5 above)
Comment (johnpierson) [reply src/SampleLinter/Rules/NoGroupsLinterRule.cs]: Hmm, have some guidance for that?

File: src/SampleLinter/MigrationSettingsView.xaml (hunk 8 above)
Comment (johnpierson) [reply src/SampleLinter/MigrationSettingsView.xaml]: my mistake

File: src/SampleLinter/Rules/NoGroupsLinterRule.cs (hunk 5 above)
Comment (BogdanZavu) [reply src/SampleLinter/Rules/NoGroupsLinterRule.cs]: Yes, checkout Properties.Resources.LocalStringResult definition in SampleLibraryUI project

**Comment by BogdanZavu** [comment]
@johnpierson just a few minor things and then LGTM!
I believe we do not need all those dependencies right ? Let's keep only what we need in linter cs proj
And I think the build result should produce a ready to consume package similar to "dynamo_viewExtension\Sample View Extension\"

**Comment by lillismith** [comment]
@johnpierson @BogdanZavu  can we wrap this up before we talk about it at Demos tomorrow?

**Comment by johnpierson** [comment]
The one thing I am missing still is the "produce a ready to consume package" part as I cannot find that in the samples

**Comment by lillismith** [comment]
@johnpierson  where do you put what to make "Sample Linter"  show up as a graph type?

**Comment by BogdanZavu** [comment]
> The one thing I am missing still is the "produce a ready to consume package" part as I cannot find that in the samples

Hey @johnpierson take a look at the build result of the whole solution. You will see that it has created this folder as example 
...\DynamoSamples\dynamo_viewExtension\Sample View Extension for a view extension.
I see that we also already have ..\DynamoSamples\dynamo_linter\Sample Linter
So the only thing that we are actually missing from there is the bin folder with the actual SampleLinter.dll

**Review by BogdanZavu — APPROVED** [review]

**Comment by BogdanZavu** [comment]
Merging this, we can add further ideas and improvements later.
